test/binaryen.js/copy-expression.js.txt

```diff
@@ -0,0 +1,37 @@
+// beginning a Binaryen API trace
+#include <math.h>
+#include <map>
+#include "binaryen-c.h"
+int main() {
+  std::map<size_t, BinaryenType> types;
+  std::map<size_t, BinaryenExpressionRef> expressions;
+  std::map<size_t, BinaryenFunctionRef> functions;
+  std::map<size_t, BinaryenGlobalRef> globals;
+  std::map<size_t, BinaryenEventRef> events;
+  std::map<size_t, BinaryenExportRef> exports;
+  std::map<size_t, RelooperBlockRef> relooperBlocks;
+  std::map<size_t, ExpressionRunnerRef> expressionRunners;
+  BinaryenModuleRef the_module = NULL;
+  RelooperRef the_relooper = NULL;
+  the_module = BinaryenModuleCreate();
+  expressions[size_t(NULL)] = BinaryenExpressionRef(NULL);
+  expressions[1] = BinaryenLocalGet(the_module, 0, BinaryenTypeInt32());
+  expressions[2] = BinaryenConst(the_module, BinaryenLiteralInt32(1));
+  expressions[3] = BinaryenConst(the_module, BinaryenLiteralInt32(2));
+  expressions[4] = BinaryenIf(the_module, expressions[1], expressions[2], expressions[3]);
+  {
+    BinaryenExpressionRef children[] = { expressions[4] };
+    expressions[5] = BinaryenBlock(the_module, NULL, children, 1, BinaryenTypeInt32());
+  }
+  expressions[6] = BinaryenExpressionCopy(expressions[5], theModule);
+  BinaryenBlockGetChild(expressions[5], 0);
+  BinaryenBlockGetChild(expressions[6], 0);
+  BinaryenIfGetCondition(expressions[4]);
+  BinaryenIfGetCondition(expressions[0]); // this is bad
```

Comment:
I might be mistaken, but aren't these because that we don't trace `BinaryenBlockGetChild`'s return value? Weren't we supposed to use something like [`traceExpression`](https://github.com/WebAssembly/binaryen/blob/f23bf9a4cd45b7ed03805f217f6c4f645a513822/src/binaryen-c.cpp#L329-L348), as in other C APIs..?

Reply:
We typically trace expression creation, i.e. inserting `expressions[x] = BinaryenBlock(...)`, while getter APIs like `BinaryenBlockGetChild` just operate on the existing set of expressions. When these expressions are passed back to a C-API, their pointer value becomes replaced with `expression[x]`.

The problem here is that `ExpressionManipulator::copy` is a black box not emitting those `expression[x] = ...`es, especially when recursively cloning sub-expressions, so we get `0`s there for every expression not in the set. Hence wondering if we need to duplicate code from `ExpressionManipulator::copy` to a C-API variant that includes tracing or if there's a better way.